Recover TGeoManager::fIsOutside for transf. tracks

benedikt-voelkel opened the pull request:
When changing from one engine to the other, the TGeoManager's state
flagging if the geometry has been left might be still set to
fIsOutside == true.
That can be a remnant from the transportation of the former engine. This
state is now cached when tracks are transferred and recovered when
tracks are picked up for tracking again.

TVirtualMC can request the recovery of a geometry state via calling
TVirtualMC::TryRestoreGeometryState(Int_t trackId).
Every further touch of the TGeoManager/TGeoNavigator within the
requesting engine is done in its responsibility.

Commit:
Recover TGeoManager::fIsOutside for transf. tracks

When changing from one engine to the other, the TGeoManager's state
flagging if the geometry has been left might be still set to
fIsOutside == true.
That can be a remnant from the transportation of the former engine. This
state is now cached when tracks are transferred and recovered when
tracks are picked up for tracking again.

TVirtualMC can request the recovery of a geometry state via calling
TVirtualMC::TryRestoreGeometryState(Int_t trackId).
Every further touch of the TGeoManager/TGeoNavigator within the
requesting engine is done in its responsibility.

ihrivnac commented:
Hi @benedikt-voelkel ,
Thank you for this update. I would like to suggest the following changes:
- Please remove "Try" from the function name: as the function return boolean, we have the infor whether the operation succeeded or not.
- There is no need to add the functions in TVirtualMC, as they are note supposed to be called by the user (I hope) and TGeant3/TGeant4 can do directly. Please, remove them.

Then, to test the code, we need both geant3 and geant4_vmc PRs based on the next_root_version branch. This is the case for the geant3  PR but not for geant4_vmc.

The test configuration for this development should be:
root_v6.18.00 - installed without VMC
geant4 10.05.p01
geant3 - your PR branch based on next_root_version 
geant4_vmc - your PR branch based on next_root_version 
vgm - master or the last tagged version

Commit:
Finalise geometry recovery

The recovery of the previous geometry state of a transferred track can
be requested from the TMCManager.
A VMC engine can hence
either call
TMCManager::RestoreGeometryState(Int_t trackId, ...) to request the
recovery for a specified track by its ID
or
TMCManager::RestoreGeometryState() to request the recovery of the track
which is currently set on the current VMC stack.

benedikt-voelkel commented:
Hello @ihrivnac 
I agree on your test setup and tried the the multi run based on your suggestions

`root -l -q -b load_multi.C run_multi.C`

for the 4 scenarios
1. ABSO: GEANT3 and GAPX: GEANT4
    1. 1 event with 20 primaries
    2. 20 events with 20 primaries each
2. ABSO: GEANT4 and ABSO: GEANT3
    1. 1 event with 20 primaries
    2. 20 events with 20 primaries each

and that looks good.

E.g. I see in scenarios 1./2. (i.) 20 times **Begin primary** and 20 times **Finish primary** as expected.  Scenarios 1./2. (ii.) then show me 20x20 **Begin/Finish primary**, also expected.

Finally, everything should be there now, see the respective PRs at https://github.com/vmc-project/geant4_vmc/pull/7 and https://github.com/vmc-project/geant3/pull/9

ihrivnac commented on source/src/TVirtualMC.cxx:
```diff
@@ -12,6 +12,7 @@
 
 #include "TVirtualMC.h"
 #include "TError.h"
+#include "TMCManager.h"
```
@benedikt-voelkel 
This include is now redundant, could you remove it?
